server/aws-lsp-codewhisperer/src/language-server/agenticChat/agenticChatController.ts

```diff
@@ -574,25 +575,26 @@ export class AgenticChatController implements ChatHandlers {
                 )
             }
             const remainingCharacterBudget = this.truncateRequest(currentRequestInput)
-            //  Fix the history to maintain invariants
+            //  Get and process the messages from history DB to maintain invariants for service requests
+            let messages: DbMessage[] = []
             if (currentMessage) {
-                const isHistoryValid = this.#chatHistoryDb.fixAndValidateHistory(
+                // Get preprocessed history messages from DB with size and number of character limits
+                let messages = this.#chatHistoryDb.getPreprocessedRequestHistory(
                     tabId,
                     currentMessage,
-                    conversationIdentifier ?? '',
                     remainingCharacterBudget
                 )
-                if (!isHistoryValid) {
+                if (!this.#chatHistoryDb.validateAndFixNewMessageToolResults(messages, currentMessage)) {
```

Comment:
Calling `validateAndFixNewMessageToolResults` after `getPreprocessedRequestHistory` might result in incorrect history length calculations. We can potentially add or remove characters, the former resulting in errors if we exceed the 600K character limit.

Reply:
Ideally the `isPromptCanceled` helps prevent the race conditions when adding messages to the history and we don't require this to return a boolean value and just move it to the `getPreprocessedRequestHistory`